cpp/src/connect/connectivity.cu

```diff
@@ -0,0 +1,152 @@
+#include "weak_cc.cuh"
+
+#include "utilities/graph_utils.cuh"
+#include "utilities/error_utils.h"
+#include <cugraph.h>
+#include <algo_types.h>
+
+#include <iostream>
+#include <type_traits>
+#include <cstdint>
+
+//
+/**
+ * @brief Compute connected components. 
+ * The weak version was imported from cuML.
+ * This implementation comes from [1] and solves component labeling problem in
+ * parallel on CSR-indexes based upon the vertex degree and adjacency graph.
+ *
+ * [1] Hawick, K.A et al, 2010. "Parallel graph component labelling with GPUs and CUDA"
+ *
+ * @tparam Type the numeric type of non-floating point elements
+ * @tparam TPB_X the threads to use per block when configuring the kernel
+ * @param graph input graph; assumed undirected for weakly CC [in]
+ * @param labels gdf_column for the output labels [out]
+ * @param connectivity_type 0=WEAK; 1=STRONG
+ * @param stream the cuda stream
+ */
+template<typename IndexT,
+         int TPB_X = 32>
+std::enable_if_t<std::is_signed<IndexT>::value,gdf_error>
+gdf_connected_components_impl(gdf_graph *graph,
+                              gdf_column *labels,
+                              cugraph_connect_t connectivity_type,
+                              cudaStream_t stream)
+{
+  static auto row_offsets_ = [](const gdf_graph* G){
+    return static_cast<const IndexT*>(G->adjList->offsets->data);
+  };
+
+  static auto col_indices_ = [](const gdf_graph* G){
+    return static_cast<const IndexT*>(G->adjList->indices->data);
+  };
+
+  static auto nrows_ = [](const gdf_graph* G){
+    return G->adjList->offsets->size - 1;
+  };
+
+  static auto nnz_ = [](const gdf_graph* G){
+    return G->adjList->indices->size;
+  };
+
+
+  GDF_REQUIRE(graph != nullptr, GDF_INVALID_API_CALL);
+    
+  GDF_REQUIRE(graph->adjList != nullptr, GDF_INVALID_API_CALL);
+    
+  GDF_REQUIRE(row_offsets_(graph) != nullptr, GDF_INVALID_API_CALL);
+
+  GDF_REQUIRE(col_indices_(graph) != nullptr, GDF_INVALID_API_CALL);
+  
+  GDF_REQUIRE(labels != nullptr, GDF_INVALID_API_CALL);
+  
+  GDF_REQUIRE(labels->data != nullptr, GDF_INVALID_API_CALL);
+  
+  auto type_id = graph->adjList->offsets->dtype;
+  GDF_REQUIRE( type_id == GDF_INT32 || type_id == GDF_INT64, GDF_UNSUPPORTED_DTYPE);
+  
+  GDF_REQUIRE( type_id == graph->adjList->indices->dtype, GDF_UNSUPPORTED_DTYPE);
+  
+  //TODO: relax this requirement:
```

Comment:
Actually, we may not relax this, we need to design the API considering the extreme case (for correctness), and in the extreme case, the number of connected components == the number of vertices.

Reply:
Actually, yes we can. In the future we may have the labels be of a different type than IndexT used for row_offsets and col_ind. There's no reason we may _not_ do that. In which case, the requirement _below_ the comment would _not_ hold.

Reply:
OK, no further complaint if this does not mean we can use data type with less precision for this.